<?xml version="1.0"?>

<launch>
	<!-- ================================== PLAY ROSBAG ============================================== -->

	<!-- <node pkg="rosbag" type="play" name="rosbag" args="/home/mfehr/experimental_ws/maplab/gonzen_bags/gonzen_run_14_2019-02-15-15-36-01.bag -q -d 5"/> -->
	<!-- <node pkg="rosbag" type="play" name="rosbag" args="/home/mfehr/experimental_ws/maplab/gonzen_bags/gonzen_rsl_network_dataset_2019-02-15-11-26-42.bag -q -d 5"/> -->
	<!-- <node pkg="rosbag" type="play" name="rosbag" args="/home/mfehr/experimental_ws/maplab/gonzen_bags/gonzen_run_2_2019-02-13-15-52-12.bag -q -d 5"/> -->

	<!-- <node pkg="rosbag" type="play" name="rosbag" args="/media/mfehr/SSD_1TB/darpa_gonzen/new/run_4_2019-03-11-09-57-31.bag -q -d 5 -r 1.5"/> -->
	<!-- <node pkg="rosbag" type="play" name="rosbag" args="/media/mfehr/SSD_1TB/darpa_gonzen/new/run_6_2019-03-11-11-08-03.bag -q -d 5 -r 1.5"/> -->

  <!-- <node pkg="rosbag" type="play" name="rosbag" args="/media/mfehr/SSD_1TB/SMB_Tunnel/SR_2/SAFETY_RESEARCH_MISSION2_2019-08-21-12-27-54_0.bag -q -d 5 -s 0 -r 2.0"/> -->
  <node pkg="rosbag" type="play" name="rosbag" args="/media/mfehr/SSD_1TB/SMB_Tunnel/SR_2/SAFETY_RESEARCH_MISSION2_2019-08-21-13-06-42_1.bag -q -d 5 -s 0 -r 2.0"/>


	<!-- ================================== OKVIS ================================================== -->

	<group ns="okvis">
		<node name="okvis_node" pkg="okvis_ros" type="okvis_node" output="screen" required="true" args="--alsologtostderr --v=3">
			<param name="config_filename" value="$(find maplab_node)/share/example_calibrations/armadillo-okvis.yaml"/>
			<remap from="/camera0" to="/VersaVIS/cam0/image_raw"/>
			<remap from="/imu" to="/VersaVIS/imu"/>
			<!-- remove okvis tf frames from tf -->
			<remap from="/tf" to="/tf_okvis"/>
		</node>
	</group>

	<!-- ===================================== MAPLAB NODE =========================================== -->

	<group ns="maplab_node">
		<remap from="/odometry/maplab_odom_T_M_I" to="/okvis/okvis_node/okvis_maplab_odometry"/>

		<!-- SENSORS -->
		<arg name="sensor_calibration_file" default="$(find maplab_node)/share/example_calibrations/armadillo.yaml"/>
		<!-- only needed if there are multiple ncameras/lidars/imus/odometries in the calibration file. -->
		<arg name="selected_ncamera_sensor_id" default=""/>
		<arg name="selected_imu_sensor_id" default=""/>
		<arg name="selected_lidar_sensor_id" default=""/>
		<arg name="selected_odometry_6dof_sensor_id" default=""/>

		<!-- DATA SOURCE -->
		<arg name="datasource_type" default="rostopic"/>
		<arg name="datasource_rosbag" default=""/>
		<arg name="imu_to_camera_time_offset_ns" default="19779092"/>

		<!-- DATA PREPROCESSING -->
		<arg name="image_apply_clahe_histogram_equalization" default="false"/>

		<!-- LOCALIZATION BASED ON VISION -->
		<arg name="enable_visual_localization" default="false"/>
		<arg name="visual_localization_map_folder" default=""/>

		<!-- LOCALIZATION BASED ON LIDAR -->
		<arg name="enable_lidar_localization" default="false"/>
		<arg name="lidar_localization_map_folder" default=""/>

		<!-- ONLINE MAPPING -->
		<arg name="enable_online_mapping" default="false"/>

		<!-- MAP BUILDING -->
    <arg name="map_output_folder" default="/tmp/maplab_node/smb_long"/>
    <arg name="map_save_on_shutdown" default="true"/>
		<arg name="map_overwrite_enabled" default="false"/>
		<arg name="map_compute_localization_map" default="false"/>
		<arg name="map_initialize_and_triangulate_landmarks_when_saving" default="true"/>
		<arg name="map_run_keyframing_when_saving" default="false"/>
		<arg name="map_save_every_n_sec" default="0"/>
		<arg name="map_split_map_into_submaps_when_saving_periodically" default="false"/>

		<!-- ATTACHING RESOURCES TO MAP -->
		<arg name="map_builder_save_tracking_image_as_resources" default="false"/>
		<arg name="map_builder_save_point_clouds_as_resources" default="false"/>
		<arg name="map_builder_save_point_cloud_maps_as_resources" default="false"/>

		<!-- TF -->
		<arg name="tf_map_frame" default="map"/>
		<arg name="tf_mission_frame" default="maplab_mission"/>
		<arg name="tf_imu_frame" default="maplab_imu"/>
		<arg name="tf_imu_refined_frame" default="maplab_imu_refined"/>

		<!-- DEBUGGING AND LOGGING -->
		<arg name="alsologtostderr" default="true"/>
		<arg name="colorlogtostderr" default="true"/>
		<arg name="v" default="1"/>
		<arg name="publish_debug_markers" default="true"/>

		<arg name="detection_visualize_keypoints" default="true"/>
		<arg name="feature_tracker_visualize_feature_tracks" default="true"/>
		<arg name="feature_tracker_visualize_keypoint_matches" default="true"/>
		<arg name="feature_tracker_visualize_keypoints" default="true"/>
		<arg name="feature_tracker_visualize_keypoints_individual_frames" default="true"/>

		<!-- MAPLAB_NODE -->
		<include file="$(find maplab_node)/launch/maplab_node.launch" pass_all_args="true"/>
	</group>
</launch>
